proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,553 @@
+default* without requiring action on the part of the developer. To provide this provenance we leverage
+the existing mechanisms for collecting backtraces listed above. Furthermore, we
+want to ensure that this information is available for consumption in structured
+form by the user program, to allow use by logging libraries (for instance, see
+`kaptip-raven #1
+<https://github.com/cachix/katip-raven/issues/1#issuecomment-625389463>`_),
+automatic error reporting, code analysis tools, and the like.
+
+Proposed Change Specification
+-----------------------------
+
+This proposal consists of three largely-independent parts:
+
+1. augmenting the current root of the exception hierarchy,
+   ``SomeException``, with additional metadata in the form of an
+   ``ExceptionContext``.
+2. the introduction of annotation types to capture backtraces from the
+   above-described backtrace-collection mechanisms
+3. a facility for choosing which backtrace mechanism(s)
+   should be used to collect exception provenance.
+
+We will describe these parts below.
+
+Introducing ``ExceptionContext``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+All exceptions in Haskell are currently represented by the
+``Control.Exception.SomeException`` type which captures the exception along
+with its ``Exception`` dictionary: ::
+
+    data SomeException where
+        SomeException :: forall a. (Exception a)
+                      => a -> SomeException
+
+Such a ``SomeException`` value carries all of the information needed to catch,
+identify, and display an exception, but lacks any generic facility for encoding
+relevant metadata about the exception.
+
+As a first step towards attaching backtraces to exceptions, we propose to extend
+``SomeException`` to carry an additional "context" value, consisting of a collection
+of dynamically-typed annotations. The dynamic nature of these annotations
+allows the mechanism to handle not just backtraces but any arbitrary
+context-relevant value which may be inspected by a handler for analysis and logging.
+To avoid breaking existing users of ``SomeException``, we introduce this
+context as an implicit parameter constraint:  ::
+
+    data SomeException where
+        SomeException :: forall a. (Exception a, ?exceptionContext :: ExceptionContext)
+                      => a -> SomeException
+
+    data ExceptionContext = ExceptionContext [SomeExceptionAnnotation]
+
+    -- | Contexts are merged by concatenation
+    instance Semigroup ExceptionContext
+    instance Monoid ExceptionContext
+
+Following the example of the ``Exception`` class, we propose that exception
+annotations implement a minimal typeclass which provides ``Typeable`` evidence
+and the ability to render the annotation to a ``String``: ::
+
+    data SomeExceptionAnnotation where
+        SomeExceptionAnnotation :: forall a. (ExceptionAnnotation a)
+                                => a -> SomeExceptionAnnotation
+
+    class Typeable a => ExceptionAnnotation a where
```

Comment:
`Exception` has a `Show` superclass, should we do the same for `ExceptionAnnotation` for consistency? I guess using `DefaultSignatures` for `displayExceptionAnnotation` is fine too though.

Reply:
I suppose we could add a `Show` superclass although I am of mixed feelings as to whether it is desireable. Specifically, I see two viewpoints:

 1. That `Exception` requiring `Show` was a mistake in light of the fact that we have `displayException`. In fact, I believe that we only have it today because the original design of `Exception` didn't include a `displayException` method.
 2. That `Show` complements `displayException`, serving as a machine-readable rendering while `displayException` is human-readable.

Given that not all types have law-abiding `Show` instances, my tendency is to agree more with (1). Making `Show` a superclass of `ExceptionAnnotation` will only serve to force more non-law-abiding `Show` instances while carrying little benefit.